BootloaderCommonPkg/Library/ElfLib/ElfLib.c

```diff
@@ -0,0 +1,228 @@
+/** @file
+
+  Copyright (c) 2018, Intel Corporation. All rights reserved.<BR>
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php.
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#include <Uefi/UefiBaseType.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/BootloaderCommonLib.h>
+#include <Library/MultibootLib.h>
+#include "Elf32.h"
+
+#define IS_BOOTABLE_I386_ELF(h) \
+ ( IS_ELF (h)                             && \
+  (h->E_Ident[EI_CLASS]   == ELFCLASS32)  && \
+  (h->E_Ident[EI_DATA]    == ELFDATA2LSB) && \
+  (h->E_Ident[EI_VERSION] == EV_CURRENT)  && \
+ ((h->E_Type              == ET_EXEC)     || (h->E_Type == ET_DYN)) && \
+  (h->E_Machine           == EM_386)      && \
+  (h->E_Version           == EV_CURRENT))
+
+/**
+  Check if the image is a bootable ELF image.
+
+  @param[in]  ImageAddress      Memory address of an image
+
+  @retval     TRUE              Image is a bootable ELF image
+  @retval     FALSE             Not a bootable ELF image
+**/
+BOOLEAN
+EFIAPI
+IsElfImage (
+  IN  VOID  *ImageBase
+  )
+{
+  ELF32_EHDR *Ehdr;
+
+  Ehdr = (ELF32_EHDR *) ImageBase;
+
+  return (BOOLEAN) IS_BOOTABLE_I386_ELF (Ehdr);
+}
+
+/**
+  load ELF image symbol table
+
+  @param[in]      ImageBase   Memory address of ELF image
+  @param[in]      Ehdr        ELF file header
+  @param[in,out]  Eaddr       Start address for ELF symbols
+
+  @retval         End address of loaded sysbols table
+**/
+UINT32
+LoadElfSymtab (
+  IN  VOID   *ImageBase,
+  IN  UINT32  Eaddr
+  )
+{
+  ELF32_EHDR                 *Ehdr;
+  ELF32_SHDR                 *Shdr;
+  UINT32                      Addr;
+  UINT32                      Size;
+  UINT32                      Index;
+
+  // Copy ELF section headers.
+  Ehdr = (ELF32_EHDR *) (UINTN) ImageBase;
+  if (Ehdr->E_Shnum == 0 || Ehdr->E_Shentsize) {
+    return Eaddr;
+  }
+
+  Shdr = (ELF32_SHDR *) (UINTN) Eaddr;
+  Size = Ehdr->E_Shnum * Ehdr->E_Shentsize;
+  CopyMem (Shdr, (VOID *) ((UINT32)ImageBase + Ehdr->E_Shoff), Size);
+  Eaddr += Size;
+
+  // Copy sections that are part of the symbol table.
+  for (Index = 0 ; Index < Ehdr->E_Shnum ; Index += 1) {
+    if (Shdr[Index].Sh_Type != SHT_SYMTAB) {
```

Comment:
Two questions here: If we are loading the section header table...why not load the entire thing instead of just the symbol table? If we are loading just the symbol table..why should we load it at all? Does any kernel actually depend on the symbol table being loaded?

Reply:
All necessary things for execution binary are loaded thru program header. Only symbol information needs to be loaded thru section header. If we do not use program header, we can load the entire thing thru section header.
- Code from program header
- Symbol from section header
This is a very minimal way to load executable code and symbol because program header is much simpler than section header.

I am not sure kernel depends on the symbol table, but it may be optional for debugging. If we don't want to load the information, we can strip ELF.

Someone has any idea?

Reply:
dynamic symbols could potentially be used at RT for elf.

Reply:
SHT_DYNSYM(dynamic symbols) is a smaller version of SHT_SYMTAB(symbols). SYMTAB already includes DYNSYM.

Reply:
What I mean here is that we are providing ELF section information through the multiboot interface, but only the symbol section. It could be the case that the OS depends on other section headers as well (.text/.data/etc). Should we not also load these if we are doing this?